swap/protocol_test.go

```diff
@@ -0,0 +1,318 @@
+// We then create Swap protocol peers with a MsgPipe to be able to directly write messages to each other.
+// We have the debitor send a cheque via an `EmitChequeMsg`, then the creditor "reads" (pipe) the message
+// and handles the cheque.
+func TestEmitCheque(t *testing.T) {
+	log.Debug("set up test swaps")
+	creditorSwap, testDir1 := newTestSwap(t)
+	debitorSwap, testDir2 := newTestSwap(t)
+	defer os.RemoveAll(testDir1)
+	defer os.RemoveAll(testDir2)
+
+	ctx := context.Background()
+
+	log.Debug("deploy to simulated backend")
+	testDeploy(ctx, creditorSwap.backend, creditorSwap)
+	testDeploy(ctx, debitorSwap.backend, debitorSwap)
+	creditorSwap.backend.(*backends.SimulatedBackend).Commit()
+	debitorSwap.backend.(*backends.SimulatedBackend).Commit()
+
+	log.Debug("create peer instances")
+	// create Peer instances
+	// NOTE: remember that these are peer instances representing each **a model of the remote peer** for every local node
+	// so creditor is the model of the remote mode for the debitor! (and vice versa)
+
+	// in order to be able to model as realistically as possible sending and receiving, let's use a MsgPipe
+	// a MsgPipe is a duplex read-write object, write to one end and read from the other
+
+	// create the message pipe
+	crw, drw := p2p.MsgPipe()
+	// create the creditor peer
+	cPtpPeer := p2p.NewPeer(enode.ID{}, "creditor", []p2p.Cap{})
+	cProtoPeer := protocols.NewPeer(cPtpPeer, crw, Spec)
+	// create the debitor peer
+	dPtpPeer := p2p.NewPeer(enode.ID{}, "dreditor", []p2p.Cap{})
+	dProtoPeer := protocols.NewPeer(dPtpPeer, drw, Spec)
+	// create the Swap protocol peers
+	creditor := NewPeer(cProtoPeer, debitorSwap, debitorSwap.backend, creditorSwap.owner.address, debitorSwap.owner.Contract)
+	debitor := NewPeer(dProtoPeer, creditorSwap, creditorSwap.backend, debitorSwap.owner.address, debitorSwap.owner.Contract)
+
+	// set balance artifically
+	creditorSwap.balances[debitor.ID()] = 42
+	log.Debug("balance", "balance", creditorSwap.balances[debitor.ID()])
+	// a safe check: at this point no cheques should be in the swap
+	if len(creditorSwap.cheques) != 0 {
+		t.Fatalf("Expected no cheques at creditor, but there are %d:", len(creditorSwap.cheques))
+	}
+
+	log.Debug("create a cheque")
+	var err error
+	cheque := &Cheque{
+		ChequeParams: ChequeParams{
+			Contract:    debitorSwap.owner.Contract,
+			Beneficiary: creditorSwap.owner.address,
+			Amount:      42,
+			Honey:       42,
+			Timeout:     0,
+		},
+	}
+	cheque.Sig, err = debitorSwap.signContent(cheque)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	emitMsg := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+	log.Debug("send the message with the cheque to the beneficiary")
+	go creditor.Send(ctx, emitMsg)
+
+	log.Debug("read the message on the beneficiary through the pipe")
+	msg, err := drw.ReadMsg()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	log.Debug("convert the message to our message type (simulated p2p.protocols)")
+	var wmsg protocols.WrappedMsg
```

Comment:
Because for now we can not test end to end between two nodes because we don't have the infrastructure to run a simulated blockchain between two nodes. Thus, in order to run tests end to end between two nodes, we need to help ourselves, and we need this code to be tested.

Reply:
but do you need to actually create the nodes in such a way that you bootstrap them halfway without really bringing them up? we have wrapper methods that handle bringing up nodes, connecting them and establish protocols between them. i'm not sure which added value brings to do this test in this manner